### PR TITLE
GEODE-8671: Two threads calling get and retrieve the same PdxInstance, resulting in corruption

### PR DESCRIPTION
**Pull request by jchen21**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RegionConcurrentOperationDUnitTest.java [hunk 1]
```diff
@@ -131,6 +140,49 @@ public void getOnPreLoadedRegionFromMultipleThreadsReturnSameObject() throws Exc
...
+  @Test
+  public void getOnPartitionedRegionFromMultipleThreadsReturnsDifferentPdxInstances()
+      throws Exception {
+    String regionName = getClass().getSimpleName();
+    CacheFactory cacheFactory = new CacheFactory();
+    cacheFactory.setPdxReadSerialized(true);
+    cacheRule.createCache(cacheFactory);
+    InternalCache cache = cacheRule.getCache();
+    cache.setCopyOnRead(true);
+    Region region = cache.createRegionFactory(PARTITION)
```
Comment (DonalEvans) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RegionConcurrentOperationDUnitTest.java:152]: The compiler warning here (and elsewhere in the file) can be avoided by using `Region<Object, Object>`.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RegionConcurrentOperationDUnitTest.java [hunk 2]
```diff
@@ -131,6 +140,49 @@ public void getOnPreLoadedRegionFromMultipleThreadsReturnSameObject() throws Exc
...
+            return value;
+          }
+        }
+      };
+
+      // In this test, two threads are doing gets. One thread puts the value
+      // We expect that the threads will *always* get different PdxInstance values
+      Future get1 = executorServiceRule.submit(getValue);
+      Future get2 = executorServiceRule.submit(getValue);
+      Future put = executorServiceRule.submit(() -> region.put(key, new TestValue()));
```
Comment (DonalEvans) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RegionConcurrentOperationDUnitTest.java:172]: Compiler warnings here (and elsewhere in the file) can be avoided by using `Future<Object>`.

**Review by DonalEvans — APPROVED** [review]
Looks good, just a few small code clean-up suggestions.

**Review by upthewaterspout — APPROVED** [review]
Looks good. I'm curious to see whether this has any impact on performance in other areas since we changed the general CopyHelper class. But I think that is actually the right thing to do, since PdxInstances are not safe to share between threads.

**Comment by upthewaterspout** [comment]
Minor nitpick - can you please provide a description of what you changed and why in the final commit message. I don't see one in this PR or the individual commits, but maybe I missed something. Thanks!

**Review by dschneider-pivotal — CHANGES_REQUESTED** [review]
I think the real bug here is that some methods on PdxInstance are not thread safe. They should all be thread safe. Once we make them thread safe then we should return the same PdxInstance even if copy-on-read is true just like we do for other immutable objects.
If we can identify the PdxInstance methods that are not thread safe, then we can discuss how to make them so. The original ticket mentioned toString and serialization as not being thread safe. We need a test that demonstrates what is not thread safe.

**Review by kirklund — CHANGES_REQUESTED** [review]
I recently created a new `DistributedBlackboard` Rule for dunit tests. You can then use it as an API:
```
@Rule
public DistributedBlackboard blackboard = new DistributedBlackboard();
```
I updated several dunit tests to use it including:
```
geode-core/src/distributedTest/java/org/apache/geode/internal/cache/partitioned/fixed/FixedPartitioningDUnitTest.java
```

**Review by dschneider-pivotal — APPROVED** [review]

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RegionConcurrentOperationDUnitTest.java (hunk 2 above)
Comment (kirklund) [reply geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RegionConcurrentOperationDUnitTest.java]: If the Future is not going to have a result then go with `Future<Void>`.

File: geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RegionConcurrentOperationDUnitTest.java [hunk 3]
```diff
@@ -61,9 +63,14 @@ private static DUnitBlackboard getBlackboard() {
   @Rule
   public ExecutorServiceRule executorServiceRule = new ExecutorServiceRule();
 
+  @Rule
+  public DistributedBlackboard blackboard = new DistributedBlackboard();
+
   @After
   public void tearDown() {
-    blackboard.initBlackboard();
+    if (blackboard != null) {
```
Comment (kirklund) [inline geode-core/src/distributedTest/java/org/apache/geode/internal/cache/RegionConcurrentOperationDUnitTest.java:71]: You can delete this `tearDown`. The `DistributedBlackboard` Rule handles this for you. (If it doesn't, that's a bug that I need to know about.)

**Review by kirklund — APPROVED** [review]
Looks like you have a few small changes to make but I'll give my approval now. Please consider adding unit test coverage for the changes in LocalRegion. There is an existing `LocalRegionTest` that uses Mockito. I also suggest you remove the `tearDown` since it's doing the same thing that the `DistributedBlackboard` Rule does after every test method.

**Comment by kirklund** [comment]
> Looks like you have a few small changes to make but I'll give my approval now. Please consider adding unit test coverage for the changes in LocalRegion. There is an existing `LocalRegionTest` that uses Mockito. I also suggest you remove the `tearDown` since it's doing the same thing that the `DistributedBlackboard` Rule does after every test method.

By the way, I verified that you don't need to invoke `DUnitBlackboard.initBlackboard()` in the `tearDown`. The Rule completely dereferences the instance of `DUnitBlackboard` and so it becomes available to GC. The Rule handles the initialization and cleanup of `DUnitBlackboard`.

**Pull request merged**